src/main/java/com/hazelcast/kubernetes/DnsEndpointResolver.java

```diff
@@ -17,88 +17,72 @@
         super(logger);
         this.serviceDns = serviceDns;
         this.port = port;
-        this.dirContext = dirContext;
-    }
-
-    DnsEndpointResolver(ILogger logger, String serviceDns, int port, int serviceDnsTimeout) {
-        this(logger, serviceDns, port, createDirContext(serviceDnsTimeout));
-
-    }
-
-    @SuppressWarnings("checkstyle:magicnumber")
-    private static DirContext createDirContext(int serviceDnsTimeout) {
-        Hashtable<String, String> env = new Hashtable<String, String>();
-        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.dns.DnsContextFactory");
-        env.put(Context.PROVIDER_URL, "dns:");
-        env.put("com.sun.jndi.dns.timeout.initial", String.valueOf(serviceDnsTimeout * 1000L));
-        try {
-            return new InitialDirContext(env);
-        } catch (NamingException e) {
-            throw new HazelcastException("Error while initializing DirContext", e);
-        }
+        this.serviceDnsTimeout = serviceDnsTimeout;
     }
 
     List<DiscoveryNode> resolve() {
         try {
             return lookup();
-        } catch (NameNotFoundException e) {
-            logger.warning(String.format("DNS lookup for serviceDns '%s' failed: name not found", serviceDns));
-            return Collections.emptyList();
         } catch (Exception e) {
             logger.warning(String.format("DNS lookup for serviceDns '%s' failed", serviceDns), e);
             return Collections.emptyList();
         }
     }
 
     private List<DiscoveryNode> lookup()
-            throws NamingException, UnknownHostException {
+            throws UnknownHostException, InterruptedException, ExecutionException, TimeoutException {
         Set<String> addresses = new HashSet<String>();
-        Attributes attributes = dirContext.getAttributes(serviceDns, new String[]{"SRV"});
-        Attribute srvAttribute = attributes.get("srv");
-        if (srvAttribute != null) {
-            NamingEnumeration<?> servers = srvAttribute.getAll();
-            while (servers.hasMore()) {
-                String server = (String) servers.next();
-                String serverHost = extractHost(server);
-                InetAddress address = InetAddress.getByName(serverHost);
+
+        Future<InetAddress[]> future = DNS_LOOKUP_SERVICE.submit(new Callable<InetAddress[]>() {
+            @Override
+            public InetAddress[] call() throws Exception {
+                return getAllInetAddresses();
+            }
+        });
+
+        try {
+            for (InetAddress address : future.get(serviceDnsTimeout, TimeUnit.SECONDS)) {
                 if (addresses.add(address.getHostAddress()) && logger.isFinestEnabled()) {
                     logger.finest("Found node service with address: " + address);
                 }
             }
+        } catch (TimeoutException e) {
+            // cancel DNS lookup
+            future.cancel(true);
+            throw e;
```

Comment:
I think we should not (re)throw this exception here. Usually, when you start the Hazelcast cluster the service is not ready, so in the first starting member you'll almost always see the following exception (which does not look healthy). 

```
java.util.concurrent.TimeoutException
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
        at com.hazelcast.kubernetes.DnsEndpointResolver.lookup(DnsEndpointResolver.java:77)
        at com.hazelcast.kubernetes.DnsEndpointResolver.resolve(DnsEndpointResolver.java:58)
        at com.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategy.discoverNodes(HazelcastKubernetesDiscoveryStrategy.java:135)
        at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.discoverNodes(DefaultDiscoveryService.java:71)
        at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddresses(DiscoveryJoiner.java:70)
        at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddressesForInitialJoin(DiscoveryJoiner.java:59)
        at com.hazelcast.cluster.impl.TcpIpJoiner.joinViaPossibleMembers(TcpIpJoiner.java:138)
        at com.hazelcast.cluster.impl.TcpIpJoiner.doJoin(TcpIpJoiner.java:98)
        at com.hazelcast.internal.cluster.impl.AbstractJoiner.join(AbstractJoiner.java:137)
        at com.hazelcast.instance.Node.join(Node.java:820)
        at com.hazelcast.instance.Node.start(Node.java:455)
        at com.hazelcast.instance.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:136)
        at com.hazelcast.instance.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:203)
        at com.hazelcast.instance.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:182)
        at com.hazelcast.instance.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:132)
        at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:91)
        at com.hazelcast.core.server.StartServer.main(StartServer.java:46)
```

Maybe better do something like this (or just fall through and add a comment)?

```
            logger.warning("Could not find any service for serviceDns '" + serviceDns + "' (DNS resolution timeout)");
            return Collections.emptyList();
```

Plus, we also need to catch `UnknownHostException` for the same reason.

Reply:
I added better exception handling to the resolve() method (like it was before).
For the other two possible exceptions (InterruptedException and ExecutionException), I think it is ok to log them with the complete stacktrace, because those are "unexpected" exceptions, while the TimeoutException and UnknownHostException might be "expected" for some time (like you stated).

Reply:
There is still something wrong with the exception handling. I played with it on GKE and got the following exceptions:
```
java.util.concurrent.ExecutionException: java.net.UnknownHostException: hazelcast
        at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
        at java.util.concurrent.FutureTask.get(FutureTask.java:205) ~[?:?]
        at com.hazelcast.kubernetes.DnsEndpointResolver.lookup(DnsEndpointResolver.java:83) ~[hazelcast-kubernetes-2.2.3-SNAPSHOT.jar:?]
        at com.hazelcast.kubernetes.DnsEndpointResolver.resolve(DnsEndpointResolver.java:58) [hazelcast-kubernetes-2.2.3-SNAPSHOT.jar:?]
        at com.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategy.discoverNodes(HazelcastKubernetesDiscoveryStrategy.java:135) [hazelcast-kubernetes-2.2.3-SNAPSHOT.jar:?]
        at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.discoverNodes(DefaultDiscoveryService.java:72) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddresses(DiscoveryJoiner.java:71) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.internal.cluster.impl.DiscoveryJoiner.getPossibleAddressesForInitialJoin(DiscoveryJoiner.java:60) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.internal.cluster.impl.TcpIpJoiner.joinViaPossibleMembers(TcpIpJoiner.java:136) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.internal.cluster.impl.TcpIpJoiner.doJoin(TcpIpJoiner.java:96) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.internal.cluster.impl.AbstractJoiner.join(AbstractJoiner.java:137) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.instance.impl.Node.join(Node.java:810) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.instance.impl.Node.start(Node.java:473) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.instance.impl.HazelcastInstanceImpl.<init>(HazelcastInstanceImpl.java:123) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.instance.impl.HazelcastInstanceFactory.constructHazelcastInstance(HazelcastInstanceFactory.java:211) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:190) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:128) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:91) [hazelcast-all-4.2.jar:4.2]
        at com.hazelcast.core.server.HazelcastMemberStarter.main(HazelcastMemberStarter.java:47) [hazelcast-all-4.2.jar:4.2]
Caused by: java.net.UnknownHostException: hazelcast
        at java.net.InetAddress$CachedAddresses.get(InetAddress.java:797) ~[?:?]
        at java.net.InetAddress.getAllByName0(InetAddress.java:1505) ~[?:?]
        at java.net.InetAddress.getAllByName(InetAddress.java:1364) ~[?:?]
        at java.net.InetAddress.getAllByName(InetAddress.java:1298) ~[?:?]
```

The issue is that the `UnknownHostException` is wrapped into `ExecutionException` and then we don't catch it. Could you please check and fix it?

Reply:
You are right, I fixed this by unwrapping and throwing the cause of the `ExecutionException` (if it is an `UnknownHostException`).
I also added an assertion on the logged warning and a test case for the timeout (also with assertion on the logged warning).